changelogs/unreleased/gh-9717-fix-crash-on-new-connection-drop.md

```diff
@@ -0,0 +1,3 @@
+## bugfix/core
```

Comment:
Please add a test to the CE version since this is a bug in CE.

Reply:
Done. The test crashes the over way if we yield on disconnect trigger. It crashes due to access to connection destroyed and poisoned after yield. Yet we test the path where assertion was with the patch too.

<details>
<summary>diff</summary>

```diff
diff --git a/test/box-luatest/shutdown_test.lua b/test/box-luatest/shutdown_test.lua
index be2402f03..cac0bfe53 100644
--- a/test/box-luatest/shutdown_test.lua
+++ b/test/box-luatest/shutdown_test.lua
@@ -1,5 +1,6 @@
 local server = require('luatest.server')
 local utils = require('luatest.utils')
+local justrun = require('test.justrun')
 local fio = require('fio')
 local popen = require('popen')
 local fiber = require('fiber')
@@ -74,6 +75,35 @@ g_crash.test_shutdown_during_snapshot_on_signal = function(cg)
     t.assert_equals(status.exit_code, 0)
 end
 
+-- Test shutdown when new iproto connection is accepted but
+-- not yet fully established.
+g_crash.test_shutdown_during_new_connection = function(cg)
+    local script = [[
+        local net = require('net.box')
+        local fiber = require('fiber')
+
+        local sock = 'unix/:./iproto.sock'
+        box.cfg{listen = sock}
+        local cond = fiber.cond()
+        local in_trigger = false
+        box.session.on_connect(function()
+            in_trigger = true
+            cond:signal()
+            fiber.sleep(1)
+        end)
+        net.connect(sock, {wait_connected = false})
+        while not in_trigger do
+            cond:wait()
+        end
+        os.exit()
+    ]]
+    -- Inherit environment to inherit ASAN suppressions.
+    local env = table.copy(os.environ())
+    local result = justrun.tarantool(cg.workdir, env, {'-e', script},
+                                     {nojson = true, quote_args = true})
+    t.assert_equals(result.exit_code, 0)
+end
+
 local g = t.group()
 
 g.before_each(function(cg)

```

</details>